config/yaml-mp/src/main/java/io/helidon/config/yaml/mp/YamlConfigSourceProvider.java

```diff
@@ -21,30 +21,47 @@
 import java.util.Enumeration;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
 
 import io.helidon.config.ConfigException;
+import io.helidon.config.mp.spi.MpConfigSourceProvider;
 
 import org.eclipse.microprofile.config.spi.ConfigSource;
-import org.eclipse.microprofile.config.spi.ConfigSourceProvider;
 
 /**
  * YAML config source provider for MicroProfile config that supports file {@code application.yaml}.
  * This class should not be used directly - it is loaded automatically by Java service loader.
  */
-public class YamlConfigSourceProvider implements ConfigSourceProvider {
+public class YamlConfigSourceProvider implements MpConfigSourceProvider {
+    private static final String RESOURCE_NAME = "application.yaml";
+
     @Override
     public Iterable<ConfigSource> getConfigSources(ClassLoader classLoader) {
-        Enumeration<URL> resources;
-        try {
-            resources = classLoader.getResources("application.yaml");
-        } catch (IOException e) {
-            throw new ConfigException("Failed to read resources from classpath", e);
-        }
+        return getConfigSources(classLoader, Optional.empty());
+    }
+
+    @Override
+    public Iterable<ConfigSource> getConfigSources(ClassLoader classLoader, String profile) {
+        Objects.requireNonNull(profile, "Profile must not be null");
+        return getConfigSources(classLoader, Optional.of(profile));
+    }
 
+    private Iterable<ConfigSource> getConfigSources(ClassLoader classLoader, Optional<String> profile) {
         List<ConfigSource> result = new LinkedList<>();
-        while (resources.hasMoreElements()) {
-            URL url = resources.nextElement();
-            result.add(YamlMpConfigSource.create(url));
+        if (profile.isEmpty()) {
```

Comment:
When using a profile, we need to load the source without a profile, and then also load all profile sources and combine them so that a config source with profile overrides config source without a profile.

Reply:
This is the behavior we have now in YamlMpConfigSource.classPath() and new unit tests were added to validate when a config property is overridden (config property is added in config profile files) and not overridden (config property is only on the default config and not added in config profile files).